### PR TITLE
Not returning orphan/unknown tenant volumes during docker volume ls

### LINKED ISSUE #1111: Volume created from vm1 (ESX1) inside tenant1  but is visible from vm2 (ESX2).
Setup:
```
ESX-1 has vm1, vm2, vm3
ESX-2 has vm4

vm1 and vm4 are on shared vmfs.
```

Steps:
```
1. Created a volume from VM1. 
2. Volume visible from VM2 and VM3.
3. Created a tenant T-1 and added vm1 to it.
4. Did docker volume ls from VM1. As expected, do not see any volumes.
5. Created volume vol-1 from VM1. Docker volume ls lists the volume vol-1.
5. Did docker volume ls from VM2 and vm3, do not see vol-1.
6. Did docker volume ls from vm4, able to see vol-1.  <<<<<<<<<<<<<<<<<<<< vol-1 was created by vm1 after it moved to T-1
7. Created a tenant t-2 and added vm4 to it.
8  Did docker volume ls from vm4, still see vol-1.    <<<<<<<<<<<<<<<<<<
```
Even though I see vol-1, if I try to do docker inspect vol-1, I get the message - No such volume

On ESX-2, admin cli lists vm-group as N/A for vol-1. 

Is this a bug or it is an expected behavior? 

Should we list the volume on a different vm which is on a different ESX even though that volume belongs to user created tenant?

================================================================
Steps and their output are as follows:

- Create a volume from VM-1
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vsphere --name=TestVol -o size=500mb
TestVol
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             TestVol@sharedVmfs-0
vsphere             testVol1_Post@sharedVmfs-0
root@sc-rdops-vm02-dhcp-52-237:~#



admin cli volume ls:

[root@sc2-rdops-vm01-dhcp-34-157:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume         Datastore     VM-Group  Capacity  Used  Filesystem  Policy  Disk Format  Attached-to  Access      Attach-as               Created By           Created Date
-------------  ------------  --------  --------  ----  ----------  ------  -----------  -----------  ----------  ----------------------  -------------------  ------------------------
TestVol        sharedVmfs-0  _DEFAULT  500MB     23MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM1-157-vmfs  Wed Mar 29 18:43:02 2017
testVol1_Post  sharedVmfs-0  N/A       200MB     14MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM1-157-vmfs  Wed Mar 29 17:23:43 2017

[root@sc2-rdops-vm01-dhcp-34-157:~]


Docker volume ls from VM-2  and VM-3:

root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             TestVol@sharedVmfs-0
vsphere             testVol1_Post@sharedVmfs-0
root@sc-rdops-vm02-dhcp-52-237:~#

root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             TestVol@sharedVmfs-0
vsphere             testVol1_Post@sharedVmfs-0
root@sc-rdops-vm02-dhcp-52-237:~#

Docker volume ls from VM-4:

root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             TestVol@sharedVmfs-0
vsphere             testVol1_Post@sharedVmfs-0
root@sc-rdops-vm02-dhcp-52-237:~#

```

- Created a tenant and added a VM-1 to it:

```
[root@sc2-rdops-vm01-dhcp-34-157:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group create --name=tenant1 --vm-list=ubuntu-VM1-157-vmfs
vm-group create succeeded
[root@sc2-rdops-vm01-dhcp-34-157:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group ls
Uuid                                  Name      Description                 Default_datastore  VM_list
------------------------------------  --------  --------------------------  -----------------  -------------------
11111111-1111-1111-1111-111111111111  _DEFAULT  This is a default vm-group
cd760db8-d9db-462d-b267-a8ef3a69d7a7  tenant1                                                  ubuntu-VM1-157-vmfs

[root@sc2-rdops-vm01-dhcp-34-157:~]

```

- Created a volume - TestVol_T1 from VM-1:

```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vsphere --name=TestVol_T1 -o size=500mb
TestVol_T1
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             TestVol_T1@sharedVmfs-0
vsphere             testVol1_Post@sharedVmfs-0
root@sc-rdops-vm02-dhcp-52-237:~#



[root@sc2-rdops-vm01-dhcp-34-157:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume         Datastore     VM-Group  Capacity  Used  Filesystem  Policy  Disk Format  Attached-to  Access      Attach-as               Created By           Created Date
-------------  ------------  --------  --------  ----  ----------  ------  -----------  -----------  ----------  ----------------------  -------------------  ------------------------
TestVol        sharedVmfs-0  _DEFAULT  500MB     23MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM1-157-vmfs  Wed Mar 29 18:43:02 2017
testVol1_Post  sharedVmfs-0  N/A       200MB     14MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM1-157-vmfs  Wed Mar 29 17:23:43 2017
TestVol_T1     sharedVmfs-0  tenant1   500MB     23MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM1-157-vmfs  Wed Mar 29 18:49:29 2017

[root@sc2-rdops-vm01-dhcp-34-157:~]

```

Docker volume ls from VM-2 and VM-3:

```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             TestVol@sharedVmfs-0
vsphere             testVol1_Post@sharedVmfs-0
root@sc-rdops-vm02-dhcp-52-237:~#


root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             TestVol@sharedVmfs-0
vsphere             testVol1_Post@sharedVmfs-0
root@sc-rdops-vm02-dhcp-52-237:~#

```

Docker volume ls from VM-4:
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             TestVol@sharedVmfs-0
vsphere             TestVol_T1@sharedVmfs-0         <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
vsphere             testVol1_Post@sharedVmfs-0
root@sc-rdops-vm02-dhcp-52-237:~#

```

- Created tenant-2 and added VM-4 to it:
```
[root@sc2-rdops-vm01-dhcp-41-181:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group create --name=tenant2 --vm-list=ubuntu-VM1-181-vmfs
vm-group create succeeded

```

- Did docker volume ls from VM-4:
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             TestVol_T1@sharedVmfs-0
vsphere             testVol1_Post@sharedVmfs-0
root@sc-rdops-vm02-dhcp-52-237:~#


Admin cli volume ls ESX-2:

[root@sc2-rdops-vm01-dhcp-41-181:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume         Datastore     VM-Group  Capacity  Used  Filesystem  Policy  Disk Format  Attached-to  Access      Attach-as               Created By           Created Date
-------------  ------------  --------  --------  ----  ----------  ------  -----------  -----------  ----------  ----------------------  -------------------  ------------------------
TestVol        sharedVmfs-0  _DEFAULT  500MB     23MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM1-157-vmfs  Wed Mar 29 18:43:02 2017
testVol1_Post  sharedVmfs-0  N/A       200MB     14MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM1-157-vmfs  Wed Mar 29 17:23:43 2017
TestVol_T1     sharedVmfs-0  N/A       500MB     23MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM1-157-vmfs  Wed Mar 29 18:49:29 2017

[root@sc2-rdops-vm01-dhcp-41-181:~]

```
Test steps are planned #957

Logs:
[docker-volume-vsphere_vm1.txt](https://github.com/vmware/docker-volume-vsphere/files/880157/docker-volume-vsphere_vm1.txt)
[docker-volume-vsphere_vm4.txt](https://github.com/vmware/docker-volume-vsphere/files/880155/docker-volume-vsphere_vm4.txt)
[vmdk_ops_ESX1.txt](https://github.com/vmware/docker-volume-vsphere/files/880154/vmdk_ops_ESX1.txt)
[vmdk_ops_ESX2.txt](https://github.com/vmware/docker-volume-vsphere/files/880156/vmdk_ops_ESX2.txt)

### PR DESCRIPTION
**Pull request by pshahzeb**
Orphan/unknown tenant volumes should not be returned as a list of volumes when a specific tenant is specified. (docker volume ls)

They must be listed only when the tenant regex is "*" i.e. volumes belonging to any tenant. Such regex comes when we list volumes as a part of admin cli volume ls

Fixes #1111

Testing:

VM1 on ESX1 belongs to tenant1
VM2 on ESX2 belongs to tenant2

Create a volume from VM1 and list volumes
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create -d vsphere --name vol_tenant1_esx1
vol_tenant1_esx1
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             vol_tenant1_esx1@sharedVmfs-0
```

Admin cli output on ESX1
```
[root@sc2-rdops-vm02-dhcp-74-64:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group ls
Uuid                                  Name          Description                 Default_datastore  VM_list
------------------------------------  ------------  --------------------------  -----------------  ------------
11111111-1111-1111-1111-111111111111  _DEFAULT      This is a default vm-group
c4884ced-b78d-4a33-ba11-81da39ecd5b6  tenant1_esx1                                                 ubuntu-VM0.1

[root@sc2-rdops-vm02-dhcp-74-64:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume            Datastore     VM-Group      Capacity  Used  Filesystem  Policy  Disk Format  Attached-to  Access      Attach-as               Created By    Created Date
----------------  ------------  ------------  --------  ----  ----------  ------  -----------  -----------  ----------  ----------------------  ------------  ------------------------
vol_tenant1_esx1  sharedVmfs-0  tenant1_esx1  100MB     13MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM0.1  Thu Mar 30 02:14:20 2017
```

docker volume ls on VM2. Can't see volume created by VM1.
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
```
Admin cli output on ESX2. 
```
[root@sc2-rdops-vm02-dhcp-78-203:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group ls
Uuid                                  Name          Description                 Default_datastore  VM_list
------------------------------------  ------------  --------------------------  -----------------  ------------
11111111-1111-1111-1111-111111111111  _DEFAULT      This is a default vm-group
2daf7cb1-b74c-4ef8-a617-56a001011980  tenant2_esx2                                                 ubuntu-VM0.2

[root@sc2-rdops-vm02-dhcp-78-203:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py volume ls
Volume            Datastore     VM-Group  Capacity  Used  Filesystem  Policy  Disk Format  Attached-to  Access      Attach-as               Created By    Created Date
----------------  ------------  --------  --------  ----  ----------  ------  -----------  -----------  ----------  ----------------------  ------------  ------------------------
vol_tenant1_esx1  sharedVmfs-0  N/A       100MB     13MB  ext4        N/A     thin         detached     read-write  independent_persistent  ubuntu-VM0.1  Thu Mar 30 02:14:20 2017
```

CC @ashahi1

### CODE REVIEW

File: esx_service/utils/vmdk_utils.py [hunk 1]
```diff
@@ -163,11 +163,14 @@ def get_volumes(tenant_re):
                     logging.debug("get_volumes: cannot find tenant_name for tenant_uuid=%s", sub_dir_name)
                     logging.debug("get_volumes: path=%s root=%s sub_dir_name=%s",
                                 path, root, sub_dir_name)
-                    for file_name in list_vmdks(root):
-                        volumes.append({'path': root,
-                                        'filename': file_name,
-                                        'datastore': datastore,
-                                        'tenant' : auth_data_const.ORPHAN_TENANT})
+
```
Comment (msterin) [inline esx_service/utils/vmdk_utils.py]: Line 130 looks broken to me - `path` is .../dockvol, and should be ../dockvol/__DEFAULT.

File: esx_service/utils/vmdk_utils.py [hunk 2]
```diff
@@ -163,11 +163,14 @@ def get_volumes(tenant_re):
...
+                    # return orphan volumes only in case when volumes from any tenants are asked
+                    if tenant_re == "*":
+                        for file_name in list_vmdks(root):
+                            volumes.append({'path': root,
+                                            'filename': file_name,
+                                            'datastore': datastore,
```
Comment (msterin) [inline esx_service/utils/vmdk_utils.py]: btw , L168-173 more python-ish would be something like this  
```
volumes += [ {'path': root, 'filename': file_name, 'datastore': datastore,  
                    'tenant' : auth_data_const.ORPHAN_TENANT} for  file_name in list_vmdks(root)]
```

also applies to other places in this function

File: esx_service/utils/vmdk_utils.py [hunk 3]
```diff
@@ -163,11 +163,14 @@ def get_volumes(tenant_re):
...
                     logging.debug("get_volumes: path=%s root=%s sub_dir_name=%s",
                                 path, root, sub_dir_name)
-                    for file_name in list_vmdks(root):
-                        volumes.append({'path': root,
-                                        'filename': file_name,
-                                        'datastore': datastore,
-                                        'tenant' : auth_data_const.ORPHAN_TENANT})
+
+                    # return orphan volumes only in case when volumes from any tenants are asked
+                    if tenant_re == "*":
```
Comment (msterin) [inline esx_service/utils/vmdk_utils.py]: nit: 
I think it's better to say 
```python
if tenant != '*' 
  return volumes
```
makes much less indent and easier to read.

**Comment by pshahzeb** [comment]
@msterin 
True that tenant=None version is broken and doesn't hold valid in current state where all volumes are created inside tenant directories. Getting rid of this code would also require rewriting a couple of old test cases we have in vmdkops_admin_test
I think the entire change above and refactoring with your suggestions can be separately done after the .13 cut.
CC @lipingxue

**Comment by pshahzeb** [comment]
Created #1121 for this.

**Comment by msterin** [comment]
LGTM for now - but please file an issue for the bug and the refactor.

**Comment by pshahzeb** [comment]
#1121 Issue to track refactoring of get_volumes() util and fixing the related test cases as well

**Pull request merged**